File: go-sdk/runner/exit/subscriber.go
package exit

import (
	"fmt"
	"os"
	"os/signal"
	"strings"
	"syscall"

	"github.com/go-logr/logr"

	"github.com/nats-io/nats.go"
	"github.com/spf13/viper"
	"google.golang.org/protobuf/proto"

	"github.com/konstellation-io/kai-sdk/go-sdk/internal/common"
	"github.com/konstellation-io/kai-sdk/go-sdk/internal/errors"
	kai "github.com/konstellation-io/kai-sdk/go-sdk/protos"
	"github.com/konstellation-io/kai-sdk/go-sdk/sdk"
)

const _subscriberLoggerName = "[SUBSCRIBER]"

func (er *Runner) getLoggerWithName() logr.Logger {
	return er.sdk.Logger.WithName(_subscriberLoggerName)
}

func (er *Runner) startSubscriber() {
	inputSubjects := viper.GetStringSlice(common.ConfigNatsInputsKey)

	if len(inputSubjects) == 0 {
		er.getLoggerWithName().Info("Undefined input subjects")
		os.Exit(1)
	}

	subscriptions := make([]*nats.Subscription, 0, len(inputSubjects))

	for _, subject := range inputSubjects {
		consumerName := fmt.Sprintf("%s-%s", strings.ReplaceAll(subject, ".", "-"),
			strings.ReplaceAll(strings.ReplaceAll(er.sdk.Metadata.GetProcess(), ".", "-"), " ", "-"))

		er.getLoggerWithName().V(1).Info(fmt.Sprintf("Subscribing to subject %s with queue group %s", subject, consumerName))

		s, err := er.jetstream.QueueSubscribe(
			subject,
			consumerName,
			er.processMessage,
			nats.DeliverNew(),
			nats.Durable(consumerName),
			nats.ManualAck(),
			nats.AckWait(viper.GetDuration(common.ConfigRunnerSubscriberAckWaitTimeKey)),
		)
		if err != nil {
			er.getLoggerWithName().Error(err, fmt.Sprintf("Error subscribing to subject %s", subject))
			os.Exit(1)
		}

		subscriptions = append(subscriptions, s)

		er.getLoggerWithName().V(1).Info(fmt.Sprintf("Listening to subject %s with queue group %s", subject, consumerName))
	}

	er.sdk.Logger.Info("Subscribed to all subjects successfully")

	// Handle sigterm and await termChan signal
	termChan := make(chan os.Signal, 1)
	signal.Notify(termChan, syscall.SIGINT, syscall.SIGTERM)
	<-termChan

	// Handle shutdown
	er.getLoggerWithName().Info("Shutdown signal received")

	er.sdk.Logger.Info("Unsubscribing from all subjects")

	for _, s := range subscriptions {
		er.getLoggerWithName().V(1).Info(fmt.Sprintf("Unsubscribing from subject %s", s.Subject))

		err := s.Unsubscribe()
		if err != nil {
			er.getLoggerWithName().Error(err, fmt.Sprintf("Error unsubscribing from the subject %s", s.Subject))
			os.Exit(1)
		}
	}

	er.getLoggerWithName().Info("Unsubscribed from all subjects")
}

func (er *Runner) processMessage(msg *nats.Msg) {
	requestMsg, err := er.newRequestMessage(msg.Data)
	if err != nil {
		errMsg := fmt.Sprintf("Error parsing msg.data coming from subject %s because is not a valid protobuf: %s",
			msg.Subject, err)
		er.processRunnerError(msg, errMsg, requestMsg.RequestId)

		return
	}

	er.getLoggerWithName().Info(fmt.Sprintf("New message received with subject %s",
		msg.Subject), requestMsg.RequestId)

	handler := er.getResponseHandler(strings.ToLower(requestMsg.FromNode))
	if handler == nil {
		errMsg := fmt.Sprintf("Error missing handler for node %q", requestMsg.FromNode)
		er.processRunnerError(msg, errMsg, requestMsg.RequestId)

		return
	}

	// Make a shallow copy of the sdk object to set inside the request msg.
	hSdk := sdk.ShallowCopyWithRequest(&er.sdk, requestMsg)

	if er.preprocessor != nil {
		err := er.preprocessor(hSdk, requestMsg.Payload)
		if err != nil {
			errMsg := fmt.Sprintf("Error in node %q executing handler preprocessor for node %q: %s",
				er.sdk.Metadata.GetProcess(), requestMsg.FromNode, err)
			er.processRunnerError(msg, errMsg, requestMsg.RequestId)

			return
		}
	}

	err = handler(hSdk, requestMsg.Payload)
	if err != nil {
		errMsg := fmt.Sprintf("Error in node %q executing handler for node %q: %s",
			er.sdk.Metadata.GetProcess(), requestMsg.FromNode, err)
		er.processRunnerError(msg, errMsg, requestMsg.RequestId)

		return
	}

	if er.postprocessor != nil {
		err := er.postprocessor(hSdk, requestMsg.Payload)
		if err != nil {
			errMsg := fmt.Sprintf("Error in node %q executing handler postprocessor for node %q: %s",
				er.sdk.Metadata.GetProcess(), requestMsg.FromNode, err)
			er.processRunnerError(msg, errMsg, requestMsg.RequestId)

			return
		}
	}

	// Tell NATS we don't need to receive the message anymore, and we are done processing it.
	ackErr := msg.Ack()
	if ackErr != nil {
		er.getLoggerWithName().Error(ackErr, errors.ErrMsgAck)
	}
}

func (er *Runner) processRunnerError(msg *nats.Msg, errMsg, requestID string) {
	ackErr := msg.Ack()
	if ackErr != nil {
		er.getLoggerWithName().Error(ackErr, errors.ErrMsgAck)
	}

	er.getLoggerWithName().V(1).Info(errMsg)
	er.publishError(requestID, errMsg)
}

func (er *Runner) newRequestMessage(data []byte) (*kai.KaiNatsMessage, error) {
	requestMsg := &kai.KaiNatsMessage{}

	var err error
	if common.IsCompressed(data) {
		data, err = common.UncompressData(data)
		if err != nil {
			er.getLoggerWithName().Error(err, "Error reading compressed message")
			return nil, err
		}
	}

	err = proto.Unmarshal(data, requestMsg)

	return requestMsg, err
}

func (er *Runner) publishError(requestID, errMsg string) {
	responseMsg := &kai.KaiNatsMessage{
		RequestId:   requestID,
		Error:       errMsg,
		FromNode:    viper.GetString(common.ConfigMetadataProcessIDKey),
		MessageType: kai.MessageType_ERROR,
	}
	er.publishResponse(responseMsg, "")
}

func (er *Runner) publishResponse(responseMsg *kai.KaiNatsMessage, channel string) {
	outputSubject := er.getOutputSubject(channel)

	outputMsg, err := proto.Marshal(responseMsg)
	if err != nil {
		er.getLoggerWithName().
			Error(err, "Error generating output result because handler result is not a serializable Protobuf")
		return
	}

	outputMsg, err = er.prepareOutputMessage(outputMsg)
	if err != nil {
		er.getLoggerWithName().Error(err, "Error preparing output message")
		return
	}

	er.getLoggerWithName().V(1).Info(fmt.Sprintf("Publishing response with subject %s", outputSubject))

	_, err = er.jetstream.Publish(outputSubject, outputMsg)
	if err != nil {
		er.getLoggerWithName().Error(err, "Error publishing output")
	}
}

func (er *Runner) getOutputSubject(channel string) string {
	outputSubject := viper.GetString(common.ConfigNatsOutputKey)
	if channel != "" {
		return fmt.Sprintf("%s.%s", outputSubject, channel)
	}

	return outputSubject
}

// prepareOutputMessage will check the length of the message and compress it if necessary.
// Fails on compressed messages bigger than the threshold.
func (er *Runner) prepareOutputMessage(msg []byte) ([]byte, error) {
	maxSize, err := er.getMaxMessageSize()
	if err != nil {
		return nil, fmt.Errorf("error getting max message size: %s", err) //nolint:goerr113 // error is wrapped
	}

	lenMsg := int64(len(msg))
	if lenMsg <= maxSize {
		return msg, nil
	}

	outMsg, err := common.CompressData(msg)
	if err != nil {
		return nil, err
	}

	lenOutMsg := int64(len(outMsg))
	if lenOutMsg > maxSize {
		er.getLoggerWithName().V(1).Info("Compressed message size %s "+
			"exceeds maximum size allowed %s", sizeInMB(lenOutMsg), sizeInMB(maxSize))
		return nil, errors.ErrMessageToBig
	}

	er.getLoggerWithName().Info(fmt.Sprintf("Message prepared with original size %s "+
		"and compressed size %s", sizeInMB(lenMsg), sizeInMB(lenOutMsg)))

	return outMsg, nil
}

func (er *Runner) getResponseHandler(subject string) Handler {
	if responseHandler, ok := er.responseHandlers[subject]; ok {
		return responseHandler
	}

	// returns the default response handler, or nil if it doesn't exist
	return er.responseHandlers["default"]
}

func (er *Runner) getMaxMessageSize() (int64, error) {
	streamInfo, err := er.jetstream.StreamInfo(viper.GetString(common.ConfigNatsStreamKey))
	if err != nil {
		return 0, fmt.Errorf("error getting stream's max message size: %w", err)
	}

	streamMaxSize := int64(streamInfo.Config.MaxMsgSize)
	serverMaxSize := er.nats.MaxPayload()

	if streamMaxSize == -1 {
		return serverMaxSize, nil
	}

	if streamMaxSize < serverMaxSize {
		return streamMaxSize, nil
	}

	return serverMaxSize, nil
}

func sizeInMB(size int64) string {
	mbSize := float32(size) / 1024 / 1024
	return fmt.Sprintf("%.1f MB", mbSize)
}

File: py-sdk/runner/runner/task/subscriber.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from datetime import timedelta
from typing import TYPE_CHECKING

import loguru
from nats.aio.msg import Msg
from nats.js.api import ConsumerConfig, DeliverPolicy
from nats.js.client import JetStreamContext
from vyper import v

from runner.common.common import Handler
from sdk.messaging.messaging_utils import is_compressed, uncompress

if TYPE_CHECKING:
    from runner.task.task_runner import TaskRunner

from runner.task.exceptions import HandlerError, NewRequestMsgError, NotValidProtobuf
from sdk.kai_nats_msg_pb2 import KaiNatsMessage

ACK_TIME = 22  # hours


@dataclass
class TaskSubscriber:
    task_runner: "TaskRunner"
    logger: loguru.Logger = field(init=False)
    subscriptions: list[JetStreamContext.PushSubscription] = field(init=False, default_factory=list)

    def __post_init__(self) -> None:
        self.logger = self.task_runner.logger.bind(context="[TASK SUBSCRIBER]")

    async def start(self) -> None:
        input_subjects = v.get("nats.inputs")
        process = self.task_runner.sdk.metadata.get_process().replace(".", "-").replace(" ", "-")

        ack_wait_time = timedelta(hours=ACK_TIME)
        if isinstance(input_subjects, str):
            input_subjects = input_subjects.replace(" ", "").split(",")
        if isinstance(input_subjects, list):
            for subject in input_subjects:
                subject_ = subject.replace(".", "-")
                consumer_name = f"{subject_}-{process}"

                self.logger.info(f"subscribing to {subject} from queue group {consumer_name}")
                try:
                    sub = await self.task_runner.js.subscribe(
                        subject=subject,
                        queue=consumer_name,
                        durable=consumer_name,
                        cb=self._process_message,
                        config=ConsumerConfig(deliver_policy=DeliverPolicy.NEW, ack_wait=ack_wait_time.total_seconds()),
                        manual_ack=True,
                    )
                except Exception as e:
                    self.logger.error(f"error subscribing to the NATS subject {subject}: {e}")
                    await self.task_runner._shutdown_handler(asyncio.get_event_loop())

                self.subscriptions.append(sub)
                self.logger.info(f"listening to {subject} from queue group {consumer_name}")
        else:
            self.logger.debug("input subjects undefined")
            await self.task_runner._shutdown_handler(asyncio.get_event_loop())

        if len(self.subscriptions) > 0:
            self.logger.info("runner successfully subscribed")

    async def _process_message(self, msg: Msg) -> None:
        self.logger.info("new message received")
        try:
            request_msg = self._new_request_msg(msg.data)
            self.task_runner.sdk.set_request_msg(request_msg)
        except Exception as e:
            await self._process_runner_error(msg, NotValidProtobuf(msg.subject, error=e))
            return

        self.logger.info(f"processing message with request_id {request_msg.request_id} and subject {msg.subject}")

        from_node = request_msg.from_node
        handler = self._get_response_handler(from_node.lower())
        to_node = self.task_runner.sdk.metadata.get_process()

        if handler is None:
            await self._process_runner_error(msg, Exception(f"no handler defined for {from_node}"))
            return

        try:
            if self.task_runner.preprocessor is not None:
                await self.task_runner.preprocessor(self.task_runner.sdk, request_msg.payload)
        except Exception as e:
            await self._process_runner_error(
                msg, HandlerError(from_node, to_node, error=e, type="handler preprocessor")
            )
            return

        try:
            await handler(self.task_runner.sdk, request_msg.payload)
        except Exception as e:
            await self._process_runner_error(msg, HandlerError(from_node, to_node, error=e))
            return

        try:
            if self.task_runner.postprocessor is not None:
                await self.task_runner.postprocessor(self.task_runner.sdk, request_msg.payload)
        except Exception as e:
            await self._process_runner_error(
                msg, HandlerError(from_node, to_node, error=e, type="handler postprocessor")
            )
            return

        try:
            await msg.ack()
        except Exception as e:
            self.logger.error(f"error acknowledging message: {e}")

    async def _process_runner_error(self, msg: Msg, error: Exception) -> None:
        error_msg = str(error)
        self.logger.info(f"publishing error message {error_msg}")

        try:
            await msg.ack()
        except Exception as e:
            self.logger.error(f"error acknowledging message: {e}")

        await self.task_runner.sdk.messaging.send_error(error_msg)

    def _new_request_msg(self, data: bytes) -> KaiNatsMessage:
        request_msg = KaiNatsMessage()

        if is_compressed(data):
            try:
                data = uncompress(data)
            except Exception as e:
                error = NewRequestMsgError(error=e)
                self.logger.error(f"{error}")
                raise error

        try:
            request_msg.ParseFromString(data)  # deserialize from bytes
        except Exception as e:
            error = NewRequestMsgError(error=e)
            self.logger.error(f"{error}")
            raise error

        return request_msg

    def _get_response_handler(self, subject: str) -> Handler | None:
        if subject in self.task_runner.response_handlers:
            return self.task_runner.response_handlers[subject]

        return (
            self.task_runner.response_handlers["default"] if "default" in self.task_runner.response_handlers else None
        )
